R/select_best.R

```diff
@@ -0,0 +1,75 @@
+#' Obtain the best score with parameters produced by tuning
+#'
+#' Get the best estimator that was chosen by the search, i.e. estimator which gave the highest performance score
+#' (or Parameter setting that gave the best results) on the assessment data.
+#'
+#' @param x The results of `tune_grid()` or `tune_Bayes()`.
+#' @param metric A character value. The metric depend on the results and the mode of the
+#' model. e.g. for `mode = regression` the metrics could be `rmse, rsq, mase` etc.
+#' and for `mode = classification` it could be: `accuracy, kap, recall, precision, sens, spec, roc_auc, pr_auc, average_precision`
+#' etc. (See
+#'   \url{https://tidymodels.github.io/yardstick/articles/metric-types.html} for more details.)
+#' OR a user defined metric for performance.
+#' @param n_top number of top results to return.
+#' @param performance A logical value (TRUE/FALSE) to indicate if columns for the corresponding
+#' performance estimates should also be returned.
+#' @param maximize automate this later but will need an argument in the short term. Not currently in use.
+#' @return A tibble. The column names depend on the results and the mode of the
+#' model as well as the specified metric type.
+#' @examples
+#' \dontrun{
+#' initial_grid_knn <- tune_grid(ames_wflow,rs = cv_splits,grid = grid, control = grid_control(verbose = TRUE))
+#' select_best(initial_grid_knn,metric = "rsq", n_top = 2)
+#' select_best(initial_grid_knn,metric = "rmse")
+#' select_best(initial_grid_knn,metric = "rmse", performance = FALSE)
+#' }
+#' @export
+select_best <- function(x, metric=NA,n_top = 1,performance = TRUE,maximize = NULL) {
+
+  if (is.na(metric)) {
+    stop("Please specify a character value for metric to get the best score/params...")
+  }
+
+  metric = tolower(metric)
+
+  #metrics that require ascending/desc sorting
+  valid_asc_metrics <- c("rmse" , "mase" , "rpd")
```

Comment:
I would avoid hard-coding the metrics. We'll have a method (soon-ish) that will determine the direction. Most metrics should be maximized so go ahead and default it to that and remove lines 33 - 42. 

You should also check to see if it is a single character storing and use `rlang::abort()` if it does not (I'm trying to get away from `stop()` when I remember to do so).

Reply:
there are few matrics that need ascending sorting to get the best result out. i.e. I want to see the lowest `rmse` out of all. However, if I want to see `rsq` then it will have to be in the `desc` order? If this doesn't matter then I will delete this. Please advice

Reply:
or we could simply have an additional option of `sort = asc/desc` when returning the results? this way we wouldn't need it?

Reply:
Doesn't `maximize` take the place of `sort`?

Reply:
Also, I think that we'd rather avoid using almost right arguments like `sort == "asc" | sort == "ascending"` or `tolower(metric)`. I used to do that in `caret` and found it to be more trouble than it is worth. It also makes people think that all of the other arguments have shortcuts too.